File: programs/hpl-character-manager/src/state/character.rs
use {
    anchor_lang::{prelude::*, solana_program::keccak},
    hpl_toolkit::{compression::*, schema::*},
    spl_account_compression::Node,
    std::collections::HashMap,
};
/// Game character (particulary NFT) PDA Account
#[compressed_account]
pub struct CharacterSchema {
    /// The wallet that owns this character
    pub owner: Pubkey,

    /// Where this character came from
    #[chunk]
    pub source: CharacterSource,

    /// NFT being used by a HPL Service
    #[chunk]
    pub used_by: CharacterUsedBy,
}

#[compressed_account(chunk = source)]
pub enum CharacterSource {
    Wrapped {
        mint: Pubkey,
        criteria: NftWrapCriteria,
        is_compressed: bool,
    },
}

impl CharacterSource {
    pub fn id(&self) -> Pubkey {
        match self {
            Self::Wrapped { mint, .. } => *mint,
        }
    }
}

#[derive(ToSchema, AnchorSerialize, AnchorDeserialize, Clone, ToNode)]
pub enum NftWrapCriteria {
    Collection(Pubkey),
    Creator(Pubkey),
    MerkleTree(Pubkey),
}

#[derive(AnchorSerialize, AnchorDeserialize)]
pub enum DataOrHash<Data> {
    Data(Data),
    Hash([u8; 32]),
}

#[compressed_account(chunk = used_by)]
pub enum CharacterUsedBy {
    None,
    Staking {
        pool: Pubkey,
        staker: Pubkey,
        staked_at: i64,
        claimed_at: i64,
    },
    Mission {
        id: Pubkey,
        rewards: Vec<EarnedReward>,
        end_time: u64,
        rewards_collected: bool,
    },
    Guild {
        id: Pubkey,
        role: GuildRole,
        order: u8,
    },
}

impl CharacterUsedBy {
    pub fn is_used(&self) -> bool {
        match self {
            CharacterUsedBy::None => false,
            _ => true,
        }
    }

    pub fn user(&self) -> Pubkey {
        match self {
            CharacterUsedBy::None => panic!("Character is not used by anything"),
            CharacterUsedBy::Staking { pool, .. } => *pool,
            CharacterUsedBy::Mission { id, .. } => *id,
            CharacterUsedBy::Guild { id, .. } => *id,
        }
    }

    pub fn match_user(&self, user: &Pubkey) -> bool {
        &self.user() == user
    }

    pub fn match_used_by(&self, used_by: &Self) -> bool {
        match (self, used_by) {
            (Self::None, Self::None) => true,
            (Self::Staking { .. }, Self::Staking { .. }) => true,
            (Self::Mission { .. }, Self::Mission { .. }) => true,
            (Self::Guild { .. }, Self::Guild { .. }) => true,
            _ => false,
        }
    }
}

#[derive(AnchorSerialize, AnchorDeserialize, Clone, ToSchema, ToNode)]
pub struct EarnedReward {
    pub delta: u8,
    pub reward_idx: u8,
}

#[derive(AnchorSerialize, AnchorDeserialize, Clone, ToSchema, ToNode)]
pub enum GuildRole {
    Chief,
    Member,
}

File: packages/hpl-character-manager/Character.ts
import { PublicKey } from "@solana/web3.js";
import { CharacterSchema, GuildRole, NftWrapCriteria } from "./generated";
import type { Asset, CompressedData } from "./types";
import {
  boolHash,
  fetchHeliusAssets,
  getCompressedData,
  getProof,
  keccak256Hash,
} from "./utils";

export type CharacterOffchainData = {
  owner: string;
  source: {
    __kind: "Wrapped";
    params: {
      mint: string;
      criteria: {
        __kind: "Collection" | "Creator" | "MerkleTree";
        params: string;
      };
      is_compressed: boolean;
    };
  };
  used_by:
    | {
        __kind: "None";
        params: null;
      }
    | {
        __kind: "Staking";
        params: {
          pool: string;
          staker: string;
          staked_at: number;
          claimed_at: number;
        };
      }
    | {
        __kind: "Mission";
        params: {
          id: string;
          requirement: {
            __kind: "Time";
            endTime: number;
          };
          rewards: {
            delta: number;
            rewardIdx: number;
          }[];
          rewardsCollected: boolean;
        };
      }
    | {
        __kind: "Guild";
        params: {
          id: string;
          role: { __kind: "Chief" } | { __kind: "Member" };
          order: number;
        };
      };
};

export const nftWrapCriteriaHash = (criteria: NftWrapCriteria) => {
  return keccak256Hash([
    Buffer.from(criteria.__kind),
    criteria.fields[0].toBuffer(),
  ]);
};

export class HplCharacter {
  private _asset: Asset;

  constructor(
    private _schema: CharacterSchema,
    private _merkleTree: PublicKey,
    private _leafIdx: number
  ) {}

  public get owner() {
    return this._schema.owner;
  }

  public get ownerHash() {
    return this.owner.toBuffer();
  }

  public get source() {
    return this._schema.source;
  }

  public get sourceHash() {
    switch (this.source.__kind) {
      case "Wrapped":
        return keccak256Hash([
          Buffer.from("Wrapped"),
          this.source.mint.toBuffer(),
          Buffer.from(nftWrapCriteriaHash(this.source.criteria)),
          Buffer.from(boolHash(this.source.isCompressed)),
        ]);
    }
  }

  public get usedBy() {
    return this._schema.usedBy;
  }

  public get usedByHash() {
    switch (this.usedBy.__kind) {
      case "None":
        return keccak256Hash([Buffer.from("None")]);

      default:
        throw new Error(
          "Used by `" + this.usedBy.__kind + "` hash not support yet"
        );
    }
  }

  public get merkleTree() {
    return this._merkleTree;
  }

  public get leafIdx() {
    return this._leafIdx;
  }

  public get hash() {
    return keccak256Hash([
      Buffer.from("CharacterSchema"),
      this.ownerHash,
      this.sourceHash,
      this.usedByHash,
    ]);
  }

  public static parseOffchainToSchema(
    offchainData: CharacterOffchainData
  ): CharacterSchema {
    return {
      owner: new PublicKey(offchainData.owner.replace("pubkey:", "")),
      source: {
        __kind: "Wrapped",
        mint: new PublicKey(
          offchainData.source.params.mint.replace("pubkey:", "")
        ),
        criteria: {
          __kind: offchainData.source.params.criteria.__kind,
          fields: [
            new PublicKey(
              offchainData.source.params.criteria.params.replace("pubkey:", "")
            ),
          ],
        },
        isCompressed: offchainData.source.params.is_compressed,
      },
      usedBy:
        "Staking" === offchainData.used_by.__kind
          ? {
              __kind: "Staking",
              pool: new PublicKey(
                offchainData.used_by.params.pool.replace("pubkey:", "")
              ),
              staker: new PublicKey(
                offchainData.used_by.params.staker.replace("pubkey:", "")
              ),
              stakedAt: offchainData.used_by.params.staked_at,
              claimedAt: offchainData.used_by.params.claimed_at,
            }
          : "Mission" === offchainData.used_by.__kind
          ? {
              __kind: "Mission",
              id: new PublicKey(
                offchainData.used_by.params.id.replace("pubkey:", "")
              ),
              endTime: offchainData.used_by.params.requirement.endTime,
              rewards: offchainData.used_by.params.rewards,
              rewardsCollected: offchainData.used_by.params.rewardsCollected,
            }
          : "Guild" === offchainData.used_by.__kind
          ? {
              __kind: "Guild",
              id: new PublicKey(
                offchainData.used_by.params.id.replace("pubkey:", "")
              ),
              order: offchainData.used_by.params.order,
              role:
                offchainData.used_by.params.role.__kind == "Chief"
                  ? GuildRole.Chief
                  : GuildRole.Member,
            }
          : { __kind: "None" },
    };
  }

  public static fromCompressedData(
    compressedData: CompressedData<CharacterOffchainData>
  ) {
    return new HplCharacter(
      HplCharacter.parseOffchainToSchema(compressedData.parsed_data),
      new PublicKey(compressedData.tree_id),
      compressedData.leaf_idx
    );
  }

  public static async fetchWithTreeAndLeaf(
    rpcUrl: string,
    tree: PublicKey,
    leafIdx: number
  ) {
    const compressedData = await getCompressedData<CharacterOffchainData>(
      rpcUrl,
      leafIdx,
      tree
    );
    if (!compressedData) throw new Error("Character not found");
    return HplCharacter.fromCompressedData(compressedData);
  }

  public static async fetchWithWallet(rpcUrl: string, wallet: PublicKey) {
    const { result, error } = await fetch(rpcUrl, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify({
        jsonrpc: "2.0",
        id: "my-id",
        method: "getCharacters",
        params: {
          wallet: wallet.toString(),
        },
      }),
    }).then((res) => res.json());
    if (!result) {
      console.error(error);
      throw new Error("Error while fetching characters");
    }
    console.log("Result", result);
    return (result as CompressedData<CharacterOffchainData>[]).map(
      HplCharacter.fromCompressedData
    );
  }

  public async asset(rpcUrl: string): Promise<Asset> {
    if (!this._asset) {
      this._asset = await fetchHeliusAssets(rpcUrl, {
        mintList: [this.source.mint],
      }).then((x) => x[0]);
    }
    return this._asset;
  }

  public proof(rpcUrl: string) {
    return getProof(rpcUrl, this.leafIdx, this.merkleTree);
  }
}
